x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/ExceptionsHelper.java

```diff
@@ -112,7 +113,29 @@ public static String shardFailuresToErrorMsg(String jobId, ShardSearchFailure[]
         return requireNonNull(obj, paramName.getPreferredName());
     }
 
+    /**
+     * Unwrap the exception stack and return the most likely cause.
+     * This method has special handling for {@link SearchPhaseExecutionException}
+     * where it returns the cause of the first shard failure.
+     *
+     * @param t raw Throwable
+     * @return unwrapped throwable if possible
+     */
     public static Throwable unwrapCause(Throwable t) {
-       return org.elasticsearch.ExceptionsHelper.unwrapCause(t);
+        // circuit breaking exceptions are at the bottom
+        Throwable unwrappedThrowable = org.elasticsearch.ExceptionsHelper.unwrapCause(t);
+
+        if (unwrappedThrowable instanceof SearchPhaseExecutionException) {
+            SearchPhaseExecutionException searchPhaseException = (SearchPhaseExecutionException) unwrappedThrowable;
+            for (ShardSearchFailure shardFailure : searchPhaseException.shardFailures()) {
+                Throwable unwrappedShardFailure = org.elasticsearch.ExceptionsHelper.unwrapCause(shardFailure.getCause());
+
+                if (unwrappedShardFailure instanceof ElasticsearchException) {
+                    return unwrappedShardFailure;
+                }
+            }
```

Comment:
Anyway to add the rest of the shard failures as suppressed?

Reply:
Trying something like that can make the message unreadable. Note we're already logging the `SearchPhaseExecutionException`.

I think there are 2 scenarios:

1. the search fails because of some user error (e.g. time field doesn't exist in some index). In this scenario there will be a single cause and we'll bubble it up.
2. the search fails due to the cluster being unstable. Here, we will have many shard failures. I think in this scenario we'd have to look at the logs anyhow to debug what's happening.

Having said that, I'm open to suggestions if you see another way.

Reply:
I haven't explored this for search, but for indexing I use a similar method that tries to get the most important error. That means you could loop through the shared failures and get the most problematic one, e.g. `IllegalArgumentException` > `RuntimeException`, because the IAE is irrecoverable, the RE might just be a 429 (too manu requests).

Having that said: "it depends" on what you are aiming for with this